Update developer docs (#444)
department-of-veterans-affairs/vets-design-system-documentation: src/_documentation/for-developers.md

title: For developers
tags: Installation, CSS, CSS architecture
anchors:
  - anchor: Using the Design System
  - anchor: Contributing to the Design System
slug: developers
---

# Documentation for developers

<div class="va-introtext">
How to install and use VA Design System styles and components with your project.
</div>

## Using the Design System

**If you are working in the vets-website repository**, you can skip straight to the [developer documentation](https://department-of-veterans-affairs.github.io/veteran-facing-services-tools/). Otherwise, proceed below.

> How you implement VADS styles into your project depends on how your project is structured and your preferences. The easiest way to get started is by using `npm`.

Author: I don't see any real harm in it - I'll add it with the caveat that it should only be for prototyping.

> How you implement VADS styles into your project depends on how your project is structured and your preferences. The easiest way to get started is by using `npm`. We will have a direct download available in the future.

Reviewer: > We will have a direct download available in the future.

Is this true? I think we can remove this until we decide it's a thing we actually want to do.

The example above is what is used on VA.gov, but you can customize this for your project.

#### Sass functions, variables, and interactive components

If you would like to use the Sass functions, such as for [spacing](../design/spacing-units.html#using-the-spacing-tokens), and variables in your project, you can import the files from your project scss. This documentation site imports Formation’s Sass files in its [application.scss](https://github.com/department-of-veterans-affairs/vets-design-system-documentation/blob/master/src/assets/stylesheets/application.scss#L5).

### Load the Web Component library

> The Design System team is working on developing a library of reusable Web Components that can be used on any HTML page. Once loaded on the page, these can be used like normal HTML tags in JSX without the need to import any additional code.

Reviewer: ```suggestion
The Design System team is working on developing a library of reusable Web Components that can be used on any HTML page or React application.
```

> This is already handled for the `vets-website` repository. To get our Web Component library set up in a new project, here is what we recommend:

Reviewer: This section is really clear. Nicely done! :raised_hands:

</div>
</div>

> #### Modifying components

Reviewer: I don't think this is a sub-section under `CSS`, is it?

- Use the Formation [naming convention](naming)
- Do not use ID selectors

## Contributing to the Design System

> If you contribute something to the Experimental Design System, we would like for a link to that code (in `vets-website`) to be provided in the documentation.

Reviewer: ```suggestion
If you contribute something to the Experimental Design System, include a link to that code (in `vets-website`) in the documentation.
```

> The Design System currently has an ["Experimental" Design System](../experimental-design) with less oversight which VFS application teams can contribute to. In order for a component to be included in the _official_ Design System, here is what we expect to see:

Reviewer: This should be covered somewhere else. That said, I want to make sure they include a link to where the code can be found when they _do_ add the documentation in `Experimental Design`.

1. Research on why this new addition is useful/necessary.

> 1. Guidance on how to use it.

Reviewer: So....yes, but no. This is true, but doesn't belong in the developer section of this documentation.

Author: I was thinking the same thing and a bit hesitant about including it. I'll :axe: it.

1. 90%+ test coverage

> 1. Any PR should be manageable in size (less than ~500 lines of code)

Reviewer: This is the part I'd like to keep. The audience for this document is developers, and this is the developer-facing part. :+1:

- Use the Formation [naming convention](naming)
- Do not use ID selectors

## Contributing to the Design System

The two main ways for developers to contribute to the Design System are by writing new components or by modifying existing components. Regardless of which type of contribution you are making, we expect each PR to:

- have at least 90% test coverage
- have appropriate comments/documentation for functions, classes, etc.
- have Storybook stories for new features

> - minimize complexity - don't build for features which aren't in the system yet

Reviewer: ```suggestion
- minimize complexity
- include only the smallest changeset required for the feature or fix
```

### New components

If you contribute something to the Experimental Design System, include a link to that code (in `vets-website`) in the documentation.

<!-- This is commented out until we have something to link to.
In order for a component to be included in the _official_ Design System, we expect [these criteria to be met]().
-->

### Modifying existing code

PRs which make a change to the Design System should be manageable in size (less than ~500 lines of code). This is to meant to:

- Save your time as the developer

> - Help prevent unnecessary things from making their way into the Design System

Reviewer: ```suggestion
- Keep PRs tightly focused
```
